Leverage std::ranges in more places

cdumez opened the pull request:
#### 48158ba01cc1850ee7799a8d504242553f7ade0e
<pre>
Leverage std::ranges in more places
<a href="https://bugs.webkit.org/show_bug.cgi?id=292357">https://bugs.webkit.org/show_bug.cgi?id=292357</a>

Reviewed by Geoffrey Garen and Sam Weinig.

Leverage std::ranges in more places to simplify our code a bit.

* Source/WTF/benchmarks/ConditionSpeedTest.cpp:
* Source/WTF/wtf/Dominators.h:
(WTF::Dominators::IterativeDominance::computeReversePostorder):
* Source/WTF/wtf/FastMalloc.cpp:
(WTF::MallocCallTracker::dumpStats):
* Source/WTF/wtf/IndexSparseSet.h:
(WTF::OverflowHandler&gt;::sort):
* Source/WTF/wtf/InterferenceGraph.h:
(WTF::InstrumentedIterableInterferenceGraph::Iterable::Iterable):
(WTF::InstrumentedIterableInterferenceGraph::Iterable::begin const): Deleted.
(WTF::InstrumentedIterableInterferenceGraph::Iterable::end const): Deleted.
(WTF::InstrumentedIterableInterferenceGraph::operator[] const): Deleted.
* Source/WTF/wtf/ListDump.h:
(WTF::sortedListDump):
(WTF::sortedMapDump):
* Source/WTF/wtf/ListHashSet.h:
(WTF::U&gt;::ListHashSet):
* Source/WTF/wtf/Liveness.h:
(WTF::Liveness::compute):
* Source/WTF/wtf/MemoryPressureHandler.cpp:
(WTF::MemoryPressureHandler::setMemoryFootprintNotificationThresholds):
* Source/WTF/wtf/NaturalLoops.h:
(WTF::NaturalLoops::NaturalLoops):
* Source/WTF/wtf/ParkingLot.cpp:
* Source/WTF/wtf/RefTrackerMixin.cpp:
(WTF::RefTracker::logAllLiveReferences):
* Source/WTF/wtf/SortedArrayMap.h:
* Source/WTF/wtf/URL.cpp:
(WTF::differingQueryParameters):
* Source/WTF/wtf/Vector.h:
(WTF::insertInUniquedSortedVector):
* Source/WTF/wtf/text/TextBreakIterator.h:
(WTF::TextBreakIteratorCache::take):
* Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp:
(WebCore::MediaKeyStatusMap::has):
(WebCore::MediaKeyStatusMap::get):
* Source/WebCore/Modules/encryptedmedia/MediaKeys.cpp:
(WebCore::MediaKeys::hasOpenSessions const):
* Source/WebCore/Modules/fetch/FetchHeaders.cpp:
(WebCore::FetchHeaders::Iterator::next):
* Source/WebCore/Modules/indexeddb/IDBDatabase.cpp:
(WebCore::sortAndRemoveDuplicates):
* Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
(WebCore::MediaMetadata::refreshArtworkImage):
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::haveMicrophoneDevice):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
(WebCore::URLPatternUtilities::URLPatternComponent::matchSpecialSchemeProtocol const):
* Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
(WebCore::URLPatternUtilities::escapeRegexStringForCharacters):
(WebCore::URLPatternUtilities::escapePatternStringForCharacters):
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::gb18030RangesCodePoint):
(PAL::gb18030RangesPointer):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::inheritsPresentationalRole const):
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::updateAnimationsAndSendEvents):
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::processKeyframeLikeObject):
(WebCore::processPropertyIndexedKeyframes):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::ensureEffectsAreSorted):
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::actionsFromContentRuleList const):
(WebCore::ContentExtensions::applyResultsToRequest):
* Source/WebCore/contentextensions/DFA.cpp:
(WebCore::ContentExtensions::printTransitions):
* Source/WebCore/contentextensions/HashableActionList.h:
(WebCore::ContentExtensions::HashableActionList::HashableActionList):
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::fontFace):
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
(WebCore::CSSCalc::generateSortedChildrenMap):
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
(WebCore::CSSNumericValue::toSum):
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/dom/DOMStringList.cpp:
(WebCore::DOMStringList::sort):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::matchingAnimations):
* Source/WebCore/dom/MutationObserver.cpp:
(WebCore::MutationObserver::notifyMutationObservers):
* Source/WebCore/dom/RadioButtonGroups.cpp:
(WebCore::RadioButtonGroup::members const):
* Source/WebCore/html/HTMLDocument.cpp:
(WebCore::HTMLDocument::supportedPropertyNames const):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::selectBestMediaSession):
(WebCore::HTMLMediaElement::updateActiveTextTrackCues):
* Source/WebCore/html/LinkIconCollector.cpp:
(WebCore::LinkIconCollector::iconsOfTypes):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::updateTickMarkValues):
* Source/WebCore/html/URLSearchParams.cpp:
(WebCore::URLSearchParams::sort):
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::pickBestImageCandidate):
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
(WebCore::MediaControlTextTrackContainerElement::updateDisplay):
* Source/WebCore/layout/formattingContexts/flex/FlexFormattingContext.cpp:
(WebCore::Layout::FlexFormattingContext::convertFlexItemsToLogicalSpace):
* Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
(WebCore::Layout::InlineContentConstrainer::prettifyRange):
* Source/WebCore/layout/formattingContexts/table/TableLayout.cpp:
(WebCore::Layout::distributeAvailableSpace):
* Source/WebCore/layout/integration/inline/InlineIteratorLogicalOrderTraversal.cpp:
(WebCore::InlineIterator::makeTextLogicalOrderCacheIfNeeded):
* Source/WebCore/loader/CrossOriginAccessControl.cpp:
(WebCore::createAccessControlPreflightRequest):
* Source/WebCore/loader/appcache/ApplicationCache.cpp:
(WebCore::ApplicationCache::setFallbackURLs):
* Source/WebCore/page/CaptionUserPreferences.cpp:
(WebCore::CaptionUserPreferences::sortedTrackListForMenu):
* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::sortedTrackListForMenu):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::selectorsForTarget):
(WebCore::ElementTargetingController::numberOfVisibilityAdjustmentRects):
* Source/WebCore/page/IntersectionObserver.cpp:
(WebCore::IntersectionObserver::IntersectionObserver):
* Source/WebCore/page/Page.cpp:
(WebCore::replaceRanges):
* Source/WebCore/page/PageColorSampler.cpp:
(WebCore::PageColorSampler::predominantColor):
* Source/WebCore/page/Performance.cpp:
(WebCore::Performance::getEntries const):
(WebCore::Performance::getEntriesByType const):
(WebCore::Performance::getEntriesByName const):
* Source/WebCore/page/PerformanceObserverEntryList.cpp:
(WebCore::PerformanceObserverEntryList::PerformanceObserverEntryList):
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::updateSnapOffsetsForScrollableArea):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::dumpProperties const):
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractRenderedText):
* Source/WebCore/platform/MIMETypeRegistry.cpp:
(WebCore::makeFixedVector):
* Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
(WebCore::ScrollSnapAnimatorState::preserveCurrentTargetForAxis):
(WebCore::chooseBoxToResnapTo):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::hasNoSession const):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::containsPersistentLicenseType):
* Source/WebCore/platform/graphics/FloatSegment.h:
(WebCore::differenceWithDilation):
* Source/WebCore/platform/graphics/GradientColorStops.h:
(WebCore::GradientColorStops::sort):
(WebCore::GradientColorStops::validateIsSorted const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::hasAudio const):
(WebCore::MediaSourcePrivate::hasVideo const):
* Source/WebCore/platform/graphics/PathUtilities.cpp:
(WebCore::addIntersectionPoints):
(WebCore::polygonsForRect):
* Source/WebCore/platform/graphics/TrackBuffer.cpp:
(WebCore::TrackBuffer::updateMinimumUpcomingPresentationTime):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::hasSelectedVideo const):
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.cpp:
(WebCore::SystemFontDatabaseCoreText::matchSystemFontUse):
* Source/WebCore/platform/ios/LegacyTileGrid.mm:
(WebCore::LegacyTileGrid::dropDistantTiles):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraintsAfterEnumeration):
* Source/WebCore/platform/mediastream/VideoPreset.h:
(WebCore::VideoPreset::sortFrameRateRanges):
* Source/WebCore/platform/mediastream/ios/AVAudioSessionCaptureDeviceManager.mm:
(WebCore::AVAudioSessionCaptureDeviceManager::setAudioCaptureDevices):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureDeviceManager.cpp:
(WebCore::hasDevice):
(WebCore::CoreAudioCaptureDeviceManager::refreshAudioCaptureDevices):
* Source/WebCore/platform/text/SuffixTree.h:
(WebCore::SuffixTree::Node::find):
* Source/WebCore/plugins/DOMPlugin.cpp:
(WebCore::makeMimeTypes):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::distributeItemIncurredIncreases):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::subdivide):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::FlexBoxIterator::next):
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::sortBorderValues):

Canonical link: <a href="https://commits.webkit.org/294387@main">https://commits.webkit.org/294387@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3bae4f02bfa97f3377b54d088d0647af9c2faac2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101650 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21318 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11633 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106808 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52284 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103690 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21626 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29818 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/77410 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34445 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104657 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16710 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91799 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57747 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16536 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9816 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51631 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/94322 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86403 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9893 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109161 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/100260 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28783 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/21196 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86386 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29144 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88000 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85951 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21872 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30706 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8417 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22946 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28711 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34000 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/123884 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28522 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34422 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31845 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30081 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren reviewed (approved):
I guess this raises the question of how we teach safe buffers that this use of iterators is safe. 

We can either use hardened iterator wrappers or treat ranges algorithms that are bounded by begin/end just like a ranged for loop (implicitly safe). 

This question doesn’t block this patch, since we were already using iterators anyway. But it’s something we should think through.

weinig commented on Source/WTF/wtf/RefTrackerMixin.cpp:
```diff
@@ -55,7 +57,7 @@ void RefTracker::logAllLiveReferences()
     Locker locker(lock);
     auto keysIterator = map.keys();
     auto keys = std::vector<void*> { keysIterator.begin(), keysIterator.end() };
```
This has nothing to do with your change, but so weird seeing std::vector here.

weinig commented on Source/WTF/wtf/SortedArrayMap.h:
```diff
@@ -161,7 +162,7 @@ template<ASCIISubset subset, typename CharacterType> constexpr std::make_unsigne
     : literal { inputLiteral }
 {
-    ASSERT_UNDER_CONSTEXPR_CONTEXT(std::all_of(literal.span().begin(), literal.span().end(), [] (char character) {
+    ASSERT_UNDER_CONSTEXPR_CONTEXT(std::ranges::all_of(literal.span(), [] (char character) {
```
While changing this line, want to fix the syntax to remove the whitespace between the [] and the (char characters)?

(https://webkit.org/code-style-guidelines/#spacing-lambda-paren)

cdumez replied:
Sure.

weinig commented on Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp:
```diff
@@ -80,7 +80,7 @@ JSC::JSValue MediaKeyStatusMap::get(JSC::JSGlobalObject& state, const BufferSour
     auto& statuses = m_session->statuses();
-    auto it = std::find_if(statuses.begin(), statuses.end(),
+    auto it = std::ranges::find_if(statuses,
         [&keyId] (auto& it) { return keyIdsMatch(it.first, keyId); });
```
If your going to make the one above one line, you got do this one too! :).

weinig commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -314,7 +315,7 @@ static String escapeRegexStringForCharacters(std::span<const CharacterType> char
 
     for (auto character : characters) {
-        if (std::find(regexEscapeCharacters.begin(), regexEscapeCharacters.end(), character) != regexEscapeCharacters.end())
+        if (std::ranges::find(regexEscapeCharacters, character) != regexEscapeCharacters.end())
```
Not sure where we are on c++23, but `std::ranges::contains` would work here.

cdumez replied:
Unfortunately std::ranges::contains is C+23, not C++20 and is not supported yet by all configurations. I have tried it recently and it built locally but not on EWS.

weinig commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2330,7 +2331,7 @@ bool AccessibilityRenderObject::inheritsPresentationalRole() const
         // based on its presentational status.
         auto& name = node->tagQName();
-        if (std::any_of(parentTags.begin(), parentTags.end(), [&name] (auto* possibleName) { return possibleName->get() == name; }))
+        if (std::ranges::any_of(parentTags, [&name] (auto* possibleName) { return possibleName->get() == name; }))
```
```suggestion
        if (std::ranges::any_of(parentTags, [&name](auto* possibleName) { return possibleName->get() == name; }))
```

weinig commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -216,7 +217,7 @@ void AnimationTimelinesController::updateAnimationsAndSendEvents(ReducedResoluti
 
         // 6. Perform a stable sort of the animation events in events to dispatch as follows.
-        std::stable_sort(events.begin(), events.end(), [] (const Ref<AnimationEventBase>& lhs, const Ref<AnimationEventBase>& rhs) {
+        std::ranges::stable_sort(events, [] (const Ref<AnimationEventBase>& lhs, const Ref<AnimationEventBase>& rhs) {
```
```suggestion
        std::ranges::stable_sort(events, [](auto& lhs, auto& rhs) {
```

weinig commented on Source/WebCore/css/CSSFontFaceSet.cpp:
```diff
@@ -505,7 +506,7 @@ CSSSegmentedFontFace* CSSFontFaceSet::fontFace(FontSelectionRequest request, con
         });
         FontSelectionAlgorithm fontSelectionAlgorithm(request, capabilities);
-        std::stable_sort(candidateFontFaces.begin(), candidateFontFaces.end(), [&fontSelectionAlgorithm](const CSSFontFace& first, const CSSFontFace& second) {
+        std::ranges::stable_sort(candidateFontFaces, [&fontSelectionAlgorithm](const CSSFontFace& first, const CSSFontFace& second) {
```
```suggestion
        std::ranges::stable_sort(candidateFontFaces, [&fontSelectionAlgorithm](auto& first, auto& second) {
```

weinig commented on Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
```diff
@@ -241,7 +242,7 @@ static Vector<ChildRepresentation, 16> generateSortedChildrenMap(const Children&
         sortedChildrenMap.append(ChildRepresentation { .index = i, .sortPriority = sortPriority(children[i]) });
 
-    std::stable_sort(sortedChildrenMap.begin(), sortedChildrenMap.end(), [](const auto& first, const auto& second) -> bool {
+    std::ranges::stable_sort(sortedChildrenMap, [](const auto& first, const auto& second) -> bool {
```
```suggestion
    std::ranges::stable_sort(sortedChildrenMap, [](auto& first, auto& second) {
```

weinig replied:
I haven't really used projections much at all, but I wonder if something like this would be more clear written as:

```c++
std::ranges::stable_sort(sortedChildrenMap, {}, &ChildRepresentation::sortPriority);
```

cdumez replied:
I was planning to look into potentially using projections in a follow-up since this is more involved and a bit riskier. I figured it was best to keep the non-trivial stuff out of the large refactoring PRs.

weinig commented on Source/WebCore/layout/formattingContexts/flex/FlexFormattingContext.cpp:
```diff
@@ -192,7 +193,7 @@ FlexLayout::LogicalFlexItems FlexFormattingContext::convertFlexItemsToLogicalSpa
             return;
 
-        std::stable_sort(flexItemList.begin(), flexItemList.end(), [&] (auto& a, auto& b) {
+        std::ranges::stable_sort(flexItemList, [&] (auto& a, auto& b) {
```
```suggestion
        std::ranges::stable_sort(flexItemList, [&](auto& a, auto& b) {
```

weinig commented on Source/WebCore/layout/formattingContexts/table/TableLayout.cpp:
```diff
@@ -159,7 +160,7 @@ static Vector<LayoutUnit> distributeAvailableSpace(const TableGrid& grid, Layout
         // <td colspan=4>#a</td><td>#b</td>
         // <td colspan=2>#c</td><td colspan=3>#d</td>
-        std::sort(spanningCells.begin(), spanningCells.end(), [&] (auto& a, auto& b) {
+        std::ranges::sort(spanningCells, [&] (auto& a, auto& b) {
```
```suggestion
        std::ranges::sort(spanningCells, [&](auto& a, auto& b) {
```

weinig commented on Source/WebCore/page/CaptionUserPreferences.cpp:
```diff
@@ -272,7 +273,7 @@ Vector<RefPtr<TextTrack>> CaptionUserPreferences::sortedTrackListForMenu(TextTra
     Collator collator;
 
-    std::sort(tracksForMenu.begin(), tracksForMenu.end(), [&] (auto& a, auto& b) {
+    std::ranges::sort(tracksForMenu, [&] (auto& a, auto& b) {
```
```suggestion
    std::ranges::sort(tracksForMenu, [&](auto& a, auto& b) {
```

weinig replied:
Given how often this specific comparison function is used, perhaps adding a TrackDisplayNameLessThan (or whatever one would want to call it) object to reuse would be useful.

weinig commented on Source/WebCore/page/CaptionUserPreferences.cpp:
```diff
@@ -316,7 +317,7 @@ Vector<RefPtr<AudioTrack>> CaptionUserPreferences::sortedTrackListForMenu(AudioT
     Collator collator;
 
-    std::sort(tracksForMenu.begin(), tracksForMenu.end(), [&] (auto& a, auto& b) {
+    std::ranges::sort(tracksForMenu, [&] (auto& a, auto& b) {
```
```suggestion
    std::ranges::sort(tracksForMenu, [&](auto& a, auto& b) {
```

weinig commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1342,7 +1343,7 @@ static void replaceRanges(Page& page, const Vector<FindReplacementRange>& ranges
     // so that we're consistent with our backwards iteration behavior per editing container when replacing text.
     auto containerNodesInOrderOfReplacement = copyToVector(rangesByContainerNode.keys());
-    std::sort(containerNodesInOrderOfReplacement.begin(), containerNodesInOrderOfReplacement.end(), [frameToTraversalIndexMap] (auto& firstNode, auto& secondNode) {
+    std::ranges::sort(containerNodesInOrderOfReplacement, [frameToTraversalIndexMap] (auto& firstNode, auto& secondNode) {
```
```suggestion
    std::ranges::sort(containerNodesInOrderOfReplacement, [frameToTraversalIndexMap](auto& firstNode, auto& secondNode) {
```

weinig commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -609,7 +610,7 @@ RenderedText extractRenderedText(Element& element)
 
-    std::sort(allTokensAndOffsets.begin(), allTokensAndOffsets.end(), [&](auto& a, auto& b) {
+    std::ranges::sort(allTokensAndOffsets, [&](auto& a, auto& b) {
         return ascendingOrder ? a.offset < b.offset : a.offset > b.offset;
```
I would probably pull the ascendingOrder check out of the lambda and do this as:

```c++
if (ascendingOrder)
    std::ranges::sort(allTokensAndOffsets, [&](auto& a, auto& b) { return a.offset < b.offset; });
else
    std::ranges::sort(allTokensAndOffsets, [&](auto& a, auto& b) { return a.offset > b.offset; });
```

weinig commented on Source/WebCore/platform/MIMETypeRegistry.cpp:
```diff
@@ -156,7 +157,7 @@ constexpr ComparableCaseFoldingASCIILiteral supportedImageMIMETypeArray[] = {
 {
     FixedVector<ASCIILiteral> result(std::size(array));
-    std::transform(std::begin(array), std::end(array), result.begin(), [] (auto literal) {
+    std::ranges::transform(array, result.begin(), [] (auto literal) {
```
```suggestion
    std::ranges::transform(array, result.begin(), [](auto literal) {
```

weinig commented on Source/WebCore/platform/graphics/FloatSegment.h:
```diff
@@ -55,7 +56,7 @@ struct FloatSegment {
 // return differenceWithDilation({ 0, totalWidth }, intersections);
 inline Vector<FloatSegment> differenceWithDilation(FloatSegment a, Vector<FloatSegment>&& bs, float dilationAmount)
```
I don't understand what the r-value for `bs` is doing here.

cdumez replied:
It returns it at the end using `WTFMove(bs)`, I guess this was a way to avoid an out-parameter?

weinig commented on Source/WebCore/platform/graphics/GradientColorStops.h:
```diff
@@ -71,7 +72,7 @@ class GradientColorStops {
             return;
 
-        std::stable_sort(m_stops.begin(), m_stops.end(), [] (auto& a, auto& b) {
+        std::ranges::stable_sort(m_stops, [] (auto& a, auto& b) {
```
```suggestion
        std::ranges::stable_sort(m_stops, [](auto& a, auto& b) {
```

weinig commented on Source/WebCore/platform/graphics/GradientColorStops.h:
```diff
@@ -111,7 +112,7 @@ class GradientColorStops {
     bool validateIsSorted() const
     {
-        return std::is_sorted(m_stops.begin(), m_stops.end(), [] (auto& a, auto& b) {
+        return std::ranges::is_sorted(m_stops, [] (auto& a, auto& b) {
```
```suggestion
        return std::ranges::is_sorted(m_stops, [](auto& a, auto& b) {
```

weinig commented on Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
```diff
@@ -146,7 +147,7 @@ bool MediaSourcePrivate::hasAudio() const
     ASSERT(m_dispatcher->isCurrent() || Thread::mayBeGCThread());
 
-    return std::any_of(m_activeSourceBuffers.begin(), m_activeSourceBuffers.end(), [] (SourceBufferPrivate* sourceBuffer) {
+    return std::ranges::any_of(m_activeSourceBuffers, [] (SourceBufferPrivate* sourceBuffer) {
```
```suggestion
    return std::ranges::any_of(m_activeSourceBuffers, [](auto sourceBuffer) {
```

cdumez committed:
Leverage std::ranges in more places
https://bugs.webkit.org/show_bug.cgi?id=292357

Reviewed by Geoffrey Garen and Sam Weinig.

Leverage std::ranges in more places to simplify our code a bit.

* Source/WTF/benchmarks/ConditionSpeedTest.cpp:
* Source/WTF/wtf/Dominators.h:
(WTF::Dominators::IterativeDominance::computeReversePostorder):
* Source/WTF/wtf/FastMalloc.cpp:
(WTF::MallocCallTracker::dumpStats):
* Source/WTF/wtf/IndexSparseSet.h:
(WTF::OverflowHandler>::sort):
* Source/WTF/wtf/InterferenceGraph.h:
(WTF::InstrumentedIterableInterferenceGraph::Iterable::Iterable):
(WTF::InstrumentedIterableInterferenceGraph::Iterable::begin const): Deleted.
(WTF::InstrumentedIterableInterferenceGraph::Iterable::end const): Deleted.
(WTF::InstrumentedIterableInterferenceGraph::operator[] const): Deleted.
* Source/WTF/wtf/ListDump.h:
(WTF::sortedListDump):
(WTF::sortedMapDump):
* Source/WTF/wtf/ListHashSet.h:
(WTF::U>::ListHashSet):
* Source/WTF/wtf/Liveness.h:
(WTF::Liveness::compute):
* Source/WTF/wtf/MemoryPressureHandler.cpp:
(WTF::MemoryPressureHandler::setMemoryFootprintNotificationThresholds):
* Source/WTF/wtf/NaturalLoops.h:
(WTF::NaturalLoops::NaturalLoops):
* Source/WTF/wtf/ParkingLot.cpp:
* Source/WTF/wtf/RefTrackerMixin.cpp:
(WTF::RefTracker::logAllLiveReferences):
* Source/WTF/wtf/SortedArrayMap.h:
* Source/WTF/wtf/URL.cpp:
(WTF::differingQueryParameters):
* Source/WTF/wtf/Vector.h:
(WTF::insertInUniquedSortedVector):
* Source/WTF/wtf/text/TextBreakIterator.h:
(WTF::TextBreakIteratorCache::take):
* Source/WebCore/Modules/encryptedmedia/MediaKeyStatusMap.cpp:
(WebCore::MediaKeyStatusMap::has):
(WebCore::MediaKeyStatusMap::get):
* Source/WebCore/Modules/encryptedmedia/MediaKeys.cpp:
(WebCore::MediaKeys::hasOpenSessions const):
* Source/WebCore/Modules/fetch/FetchHeaders.cpp:
(WebCore::FetchHeaders::Iterator::next):
* Source/WebCore/Modules/indexeddb/IDBDatabase.cpp:
(WebCore::sortAndRemoveDuplicates):
* Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
(WebCore::MediaMetadata::refreshArtworkImage):
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::haveMicrophoneDevice):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
(WebCore::URLPatternUtilities::URLPatternComponent::matchSpecialSchemeProtocol const):
* Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
(WebCore::URLPatternUtilities::escapeRegexStringForCharacters):
(WebCore::URLPatternUtilities::escapePatternStringForCharacters):
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::gb18030RangesCodePoint):
(PAL::gb18030RangesPointer):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::inheritsPresentationalRole const):
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::updateAnimationsAndSendEvents):
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::processKeyframeLikeObject):
(WebCore::processPropertyIndexedKeyframes):
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::ensureEffectsAreSorted):
* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::actionsFromContentRuleList const):
(WebCore::ContentExtensions::applyResultsToRequest):
* Source/WebCore/contentextensions/DFA.cpp:
(WebCore::ContentExtensions::printTransitions):
* Source/WebCore/contentextensions/HashableActionList.h:
(WebCore::ContentExtensions::HashableActionList::HashableActionList):
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::fontFace):
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
(WebCore::CSSCalc::generateSortedChildrenMap):
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
(WebCore::CSSNumericValue::toSum):
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
(WebCore::ComputedStylePropertyMapReadOnly::entries const):
* Source/WebCore/dom/DOMStringList.cpp:
(WebCore::DOMStringList::sort):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::matchingAnimations):
* Source/WebCore/dom/MutationObserver.cpp:
(WebCore::MutationObserver::notifyMutationObservers):
* Source/WebCore/dom/RadioButtonGroups.cpp:
(WebCore::RadioButtonGroup::members const):
* Source/WebCore/html/HTMLDocument.cpp:
(WebCore::HTMLDocument::supportedPropertyNames const):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::selectBestMediaSession):
(WebCore::HTMLMediaElement::updateActiveTextTrackCues):
* Source/WebCore/html/LinkIconCollector.cpp:
(WebCore::LinkIconCollector::iconsOfTypes):
* Source/WebCore/html/RangeInputType.cpp:
(WebCore::RangeInputType::updateTickMarkValues):
* Source/WebCore/html/URLSearchParams.cpp:
(WebCore::URLSearchParams::sort):
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
(WebCore::pickBestImageCandidate):
* Source/WebCore/html/shadow/MediaControlTextTrackContainerElement.cpp:
(WebCore::MediaControlTextTrackContainerElement::updateDisplay):
* Source/WebCore/layout/formattingContexts/flex/FlexFormattingContext.cpp:
(WebCore::Layout::FlexFormattingContext::convertFlexItemsToLogicalSpace):
* Source/WebCore/layout/formattingContexts/inline/InlineContentConstrainer.cpp:
(WebCore::Layout::InlineContentConstrainer::prettifyRange):
* Source/WebCore/layout/formattingContexts/table/TableLayout.cpp:
(WebCore::Layout::distributeAvailableSpace):
* Source/WebCore/layout/integration/inline/InlineIteratorLogicalOrderTraversal.cpp:
(WebCore::InlineIterator::makeTextLogicalOrderCacheIfNeeded):
* Source/WebCore/loader/CrossOriginAccessControl.cpp:
(WebCore::createAccessControlPreflightRequest):
* Source/WebCore/loader/appcache/ApplicationCache.cpp:
(WebCore::ApplicationCache::setFallbackURLs):
* Source/WebCore/page/CaptionUserPreferences.cpp:
(WebCore::CaptionUserPreferences::sortedTrackListForMenu):
* Source/WebCore/page/CaptionUserPreferencesMediaAF.cpp:
(WebCore::CaptionUserPreferencesMediaAF::sortedTrackListForMenu):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::selectorsForTarget):
(WebCore::ElementTargetingController::numberOfVisibilityAdjustmentRects):
* Source/WebCore/page/IntersectionObserver.cpp:
(WebCore::IntersectionObserver::IntersectionObserver):
* Source/WebCore/page/Page.cpp:
(WebCore::replaceRanges):
* Source/WebCore/page/PageColorSampler.cpp:
(WebCore::PageColorSampler::predominantColor):
* Source/WebCore/page/Performance.cpp:
(WebCore::Performance::getEntries const):
(WebCore::Performance::getEntriesByType const):
(WebCore::Performance::getEntriesByName const):
* Source/WebCore/page/PerformanceObserverEntryList.cpp:
(WebCore::PerformanceObserverEntryList::PerformanceObserverEntryList):
* Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp:
(WebCore::updateSnapOffsetsForScrollableArea):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::dumpProperties const):
* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::extractRenderedText):
* Source/WebCore/platform/MIMETypeRegistry.cpp:
(WebCore::makeFixedVector):
* Source/WebCore/platform/ScrollSnapAnimatorState.cpp:
(WebCore::ScrollSnapAnimatorState::preserveCurrentTargetForAxis):
(WebCore::chooseBoxToResnapTo):
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::hasNoSession const):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
(WebCore::containsPersistentLicenseType):
* Source/WebCore/platform/graphics/FloatSegment.h:
(WebCore::differenceWithDilation):
* Source/WebCore/platform/graphics/GradientColorStops.h:
(WebCore::GradientColorStops::sort):
(WebCore::GradientColorStops::validateIsSorted const):
* Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
(WebCore::MediaSourcePrivate::hasAudio const):
(WebCore::MediaSourcePrivate::hasVideo const):
* Source/WebCore/platform/graphics/PathUtilities.cpp:
(WebCore::addIntersectionPoints):
(WebCore::polygonsForRect):
* Source/WebCore/platform/graphics/TrackBuffer.cpp:
(WebCore::TrackBuffer::updateMinimumUpcomingPresentationTime):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaSourcePrivateAVFObjC.mm:
(WebCore::MediaSourcePrivateAVFObjC::hasSelectedVideo const):
* Source/WebCore/platform/graphics/cocoa/SystemFontDatabaseCoreText.cpp:
(WebCore::SystemFontDatabaseCoreText::matchSystemFontUse):
* Source/WebCore/platform/ios/LegacyTileGrid.mm:
(WebCore::LegacyTileGrid::dropDistantTiles):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::validateRequestConstraintsAfterEnumeration):
* Source/WebCore/platform/mediastream/VideoPreset.h:
(WebCore::VideoPreset::sortFrameRateRanges):
* Source/WebCore/platform/mediastream/ios/AVAudioSessionCaptureDeviceManager.mm:
(WebCore::AVAudioSessionCaptureDeviceManager::setAudioCaptureDevices):
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureDeviceManager.cpp:
(WebCore::hasDevice):
(WebCore::CoreAudioCaptureDeviceManager::refreshAudioCaptureDevices):
* Source/WebCore/platform/text/SuffixTree.h:
(WebCore::SuffixTree::Node::find):
* Source/WebCore/plugins/DOMPlugin.cpp:
(WebCore::makeMimeTypes):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::distributeItemIncurredIncreases):
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::subdivide):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::FlexBoxIterator::next):
* Source/WebCore/rendering/RenderTableCell.cpp:
(WebCore::RenderTableCell::sortBorderValues):

Canonical link: https://commits.webkit.org/294387@main
